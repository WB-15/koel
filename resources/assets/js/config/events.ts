export type EventName =
  'KOEL_READY'
  | 'SONG_STARTED'
  | 'LOAD_MAIN_CONTENT'
  | 'LOG_OUT'
  | 'TOGGLE_SIDEBAR'
  | 'TOGGLE_SEARCH_FORM'
  | 'SHOW_OVERLAY'
  | 'HIDE_OVERLAY'
  | 'FOCUS_SEARCH_FIELD'
  | 'PLAY_YOUTUBE_VIDEO'
  | 'INIT_EQUALIZER'
  | 'TOGGLE_VISUALIZER'
  | 'SET_SELECTED_SONGS'
  | 'SEARCH_KEYWORDS_CHANGED'
  | 'SONG_CONTEXT_MENU_REQUESTED'
  | 'ALBUM_CONTEXT_MENU_REQUESTED'
  | 'ARTIST_CONTEXT_MENU_REQUESTED'
  | 'MODAL_SHOW_CREATE_SMART_PLAYLIST_FORM'
  | 'MODAL_SHOW_EDIT_SMART_PLAYLIST_FORM'
  | 'MODAL_SHOW_ADD_USER_FORM'
  | 'MODAL_SHOW_EDIT_USER_FORM'
  | 'MODAL_SHOW_EDIT_SONG_FORM'
  | 'MODAL_SHOW_ABOUT_KOEL'
  | 'PLAYLIST_DELETE'
  | 'SMART_PLAYLIST_UPDATED'

  // upload-related
  | 'SONG_UPLOADED'
  | 'UPLOAD_QUEUE_FINISHED'

  // socket events
  | 'SOCKET_TOGGLE_PLAYBACK'
  | 'SOCKET_TOGGLE_FAVORITE'
  | 'SOCKET_PLAY_NEXT'
  | 'SOCKET_PLAY_PREV'
  | 'SOCKET_PLAYBACK_STOPPED'
  | 'SOCKET_GET_STATUS'
  | 'SOCKET_STATUS'
  | 'SOCKET_GET_CURRENT_SONG'
  | 'SOCKET_SONG'
  | 'SOCKET_SET_VOLUME'
  | 'SOCKET_VOLUME_CHANGED'
