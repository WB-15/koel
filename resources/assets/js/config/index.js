export default {
  unknownCover: (typeof window !== 'undefined' ? window.location.href : '/') + 'public/img/covers/unknown-album.png',
  appTitle: 'Koel'
}
