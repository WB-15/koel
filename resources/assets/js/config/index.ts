export * from './app'
export * from './events'
export * from './upload.types'
export * from './valid-media-mime-types'
