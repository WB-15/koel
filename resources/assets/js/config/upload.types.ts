export type UploadStatus =
  | 'Ready'
  | 'Uploading'
  | 'Uploaded'
  | 'Canceled'
  | 'Errored'

export interface UploadFile {
  id: string
  file: File
  status: UploadStatus
  name: string
  progress: number
  message?: string
}
