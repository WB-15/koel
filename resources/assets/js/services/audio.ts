export const audio = {
  context: null as AudioContext | null,
  source: null as MediaElementAudioSourceNode | null,
  element: null as HTMLMediaElement | null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext (): AudioContext {
    return this.context!
  },

  getSource (): MediaElementAudioSourceNode {
    return this.source!
  },

  getElement (): HTMLMediaElement {
    return this.element!
  }
}
