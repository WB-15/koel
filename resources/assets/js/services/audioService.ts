export const audioService = {
  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,

  init (element: HTMLMediaElement) {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context
  },

  getSource () {
    return this.source
  },

  getElement () {
    return this.element
  }
}
