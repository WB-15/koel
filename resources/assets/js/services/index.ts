export * from './httpService'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
