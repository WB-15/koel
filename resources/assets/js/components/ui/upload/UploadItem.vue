<template>
  <div :class="[file.status]" :title="file.message" class="upload-item">
    <span :style="{ width: `${file.progress}%` }" class="progress"></span>
    <span class="details">
      <span class="name">{{ file.name }}</span>
      <span class="controls">
        <Btn
          v-if="canRetry"
          data-test="retry-upload-btn"
          icon-only
          title="Retry upload"
          transparent
          unrounded
          @click="retry"
        >
          <i class="fa fa-repeat"></i>
        </Btn>
        <Btn
          v-if="canRemove"
          data-test="remove-upload-btn"
          icon-only
          title="Remove"
          transparent
          unrounded
          @click="remove"
        >
          <i class="fa fa-times"></i>
        </Btn>
      </span>
    </span>
  </div>
</template>

<script lang="ts" setup>
import { computed, defineAsyncComponent, toRefs } from 'vue'
import { UploadFile } from '@/config'
import { upload } from '@/services'

const Btn = defineAsyncComponent(() => import('@/components/ui/Btn.vue'))

const props = defineProps<{ file: UploadFile }>()
const { file } = toRefs(props)

const canRetry = computed(() => file.value.status === 'Canceled' || file.value.status === 'Errored')
const canRemove = computed(() => file.value.status !== 'Uploading') // we're not supporting cancel tokens (yet).

const remove = () => upload.remove(file.value)
const retry = () => upload.retry(file.value)
</script>

<style lang="scss" scoped>
.upload-item {
  position: relative;
  margin-bottom: 5px;
  border-radius: 3px;
  min-height: 32px;
  overflow: hidden;
  background: var(--color-bg-secondary);

  .progress {
    position: absolute;
    height: 100%;
    top: 0;
    left: 0;
    background: var(--color-highlight);
    z-index: 0;
    transition: .3s ease-out;
  }

  &.Uploaded {
    background: var(--color-green);

    .progress {
      background: transparent;
    }
  }

  &.Errored {
    background: var(--color-red);

    .progress {
      background: transparent;
    }
  }

  .details {
    z-index: 1;
    position: absolute;
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;

    .name {
      padding: 0 12px;
    }
  }

  .controls {
    display: flex;
  }

  button {
    padding: 8px 8px;
    background: rgba(0, 0, 0, .1);
  }
}
</style>
