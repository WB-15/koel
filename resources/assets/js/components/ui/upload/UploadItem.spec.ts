import { expect, it } from 'vitest'
import { UploadFile, UploadStatus } from '@/config'
import ComponentTestCase from '@/__tests__/ComponentTestCase'
import { fireEvent } from '@testing-library/vue'
import { uploadService } from '@/services'
import Btn from '@/components/ui/Btn.vue'
import UploadItem from './UploadItem.vue'

let file: UploadFile

new class extends ComponentTestCase {
  private renderComponent (status: UploadStatus) {
    file = {
      status,
      file: new File([], 'sample.mp3'),
      id: 'x-file',
      message: '',
      name: 'Sample Track',
      progress: 42
    }

    return this.render(UploadItem, {
      props: {
        file
      },
      global: {
        stubs: {
          Btn
        }
      }
    })
  }

  protected test () {
    it('renders', () => expect(this.renderComponent('Canceled').html()).toMatchSnapshot())

    it.each<[UploadStatus]>([['Canceled'], ['Errored']])('allows retrying when %s', async (status) => {
      const mock = this.mock(uploadService, 'retry')
      const { getByTitle } = this.renderComponent(status)

      await fireEvent.click(getByTitle('Retry'))

      expect(mock).toHaveBeenCalled()
    })

    it.each<[UploadStatus]>([
      ['Uploaded'],
      ['Errored'],
      ['Canceled']]
    )('allows removal if not uploading', async (status) => {
      const mock = this.mock(uploadService, 'remove')
      const { getByTitle } = this.renderComponent(status)

      await fireEvent.click(getByTitle('Remove'))

      expect(mock).toHaveBeenCalled()
    })
  }
}
