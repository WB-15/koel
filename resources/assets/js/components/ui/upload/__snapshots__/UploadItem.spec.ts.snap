// Vitest Snapshot v1

exports[`renders 1`] = `"<div class=\\"canceled upload-item\\" title=\\"\\" data-v-ad447c04=\\"\\"><span style=\\"width: 42%;\\" class=\\"progress\\" data-v-ad447c04=\\"\\"></span><span class=\\"details\\" data-v-ad447c04=\\"\\"><span class=\\"name\\" data-v-ad447c04=\\"\\">Sample Track</span><span class=\\"controls\\" data-v-ad447c04=\\"\\"><button type=\\"button\\" icon-only=\\"\\" title=\\"Retry\\" transparent=\\"\\" unrounded=\\"\\" data-v-27deb898=\\"\\" data-v-ad447c04=\\"\\"><i class=\\"fa fa-repeat\\" data-v-ad447c04=\\"\\"></i></button><button type=\\"button\\" icon-only=\\"\\" title=\\"Remove\\" transparent=\\"\\" unrounded=\\"\\" data-v-27deb898=\\"\\" data-v-ad447c04=\\"\\"><i class=\\"fa fa-times\\" data-v-ad447c04=\\"\\"></i></button></span></span></div>"`;
